cheatsheets_excluded/Content_Security_Policy_Cheat_Sheet.md

```diff
@@ -1,203 +1,199 @@
+1. `Content-Security-Policy-Report-Only` HTTP response header field. This header is used when the developer is unsure of the CSP behavior and wants to monitor it, instead of enforcing it.
 
--   Latest Revision - <https://w3c.github.io/webappsec/specs/content-security-policy/>
--   Latest Version (CSP2) - <http://www.w3.org/TR/CSP2/>
--   CSP 1.0 - <http://www.w3.org/TR/2012/CR-CSP-20121115/>
+## HTTP Headers
 
-CSP Basics
-==========
-
-CSP consists of a series of directives. CSP has also evolved over two major revisions. Most browsers support 1.0, and adoption of CSP2 has been incremental.
+The following are headers for CSP.
 
-HTTP Headers
-------------
+- `Content-Security-Policy` : W3C Spec standard header. Supported by Firefox 23+, Chrome 25+ and Opera 19+
+- `Content-Security-Policy-Report-Only` : W3C Spec standard header. Supported by Firefox 23+, Chrome 25+ and Opera 19+, whereby the policy is non-blocking ("fail open") and a report is sent to the URL designated by the `report-uri` directive. This is often used as a precursor to utilizing CSP in blocking mode ("fail closed")
+- `DO NOT` use X-Content-Security-Policy or X-WebKit-CSP. Their implementations are obsolete (since Firefox 23, Chrome 25), limited, inconsistent, and incredibly buggy.
 
-The following are headers for CSP.
+# CSP Directives
 
--   **Content-Security-Policy** : W3C Spec standard header. Supported by Firefox 23+, Chrome 25+ and Opera 19+
--   **Content-Security-Policy-Report-Only** : W3C Spec standard header. Supported by Firefox 23+, Chrome 25+ and Opera 19+, whereby the policy is non-blocking ("fail open") and a report is sent to the URL designated by the **report-uri** directive. This is often used as a precursor to utilizing CSP in blocking mode ("fail closed")
--   **DO NOT** use X-Content-Security-Policy or X-WebKit-CSP. Their implementations are obsolete (since Firefox 23, Chrome 25), limited, inconsistent, and incredibly buggy.
+Multiple types of directives exist that allow the developer to granularly control the flow of the policies.
 
-Directives
-----------
+## Fetch Directives
 
-The following is a listing of directives, and a brief description.
+Fetch directives tell the browser the locations to trust and load resources from.
 
-### CSP 1.0 Spec
+- `default-src` is a fallback directive for the other fetch directives. Directives that are specified have no inheritance, yet directives that are not specified will fall back to the value of `default-src`.
+- `child-src` allows the developer to control nested browsing contexts and worker execution contexts.
+  - `frame-src` specifies the URLs which can be loaded into nested browsing contexts (*e.g.* `<iframe>`).
+  - `worker-src` specifies the URLs which can be loaded as worker, sharedworker, or serviceworker. Fallback's on `script-src` too.
+- `connect-src` provides control over fetch requests, XHR, eventsource, beacon and websockets connections.
```

Comment:
Specify that as WebSocket do not have SOP restriction then this directive is important (or using `default-src`).

My source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/connect-src

Attack: https://www.christian-schneider.net/CrossSiteWebSocketHijacking.html

Reply:
In your source nothing was mentioned on this matter.
Find below 2 resources on the above attack using CSP:
[Saving websites from websockets SOP](https://gist.github.com/subudeepak/9897212#websockets---help-me-from-this-nightmare-please)
[content-security-policy](https://content-security-policy.com/)

Reply:
I have added the link to the attack link to show an example of exploitation of the non-limitation of the WebSocket by SOP.

Your link *Saving websites from websockets SOP* is awesome and should be added in the `connect-src` where we can be add a note about the fact that as WebSocket are not limited by the SOP, and thus the `connect-src` (or the `default-src` for fallback) directive is important.

What do you think?

Reply:
We can do that, of course. I'll see a way to organize this in a way to keep it readable to the user.

Reply:
Cool thank you 👍